File: frontend/src/lib/servertime.js
/**
 * Dark War Survival server time.
 *
 * The game announces everything against one clock, two hours behind UTC.
 * The alliance's anchor is 00:00 ST = 11:00 KST, which says the same thing:
 * KST is UTC+9, so ST is UTC+9-11 = UTC-2.
 *
 * The Etc zones invert the sign, so "Etc/GMT+2" is UTC-02:00. It has no
 * daylight saving, matching the game.
 */

export const SERVER_TZ = 'Etc/GMT+2'

const opts = (extra) => ({ timeZone: SERVER_TZ, ...extra })

/** "13:30 ST" — no date, for pairing beside a local time on the same line. */
export const fmtServerTime = (iso) =>
  `${new Intl.DateTimeFormat(undefined, opts({ hour: '2-digit', minute: '2-digit', hour12: false })).format(new Date(iso))} ST`

/** "Sat 5 Sep, 13:30 ST" — when the ST date can differ from the local one. */
export const fmtServerFull = (iso) =>
  `${new Intl.DateTimeFormat(undefined, opts({
    weekday: 'short', day: 'numeric', month: 'short',
    hour: '2-digit', minute: '2-digit', hour12: false,
  })).format(new Date(iso))} ST`

/**
 * True when a moment falls on a different calendar day in server time than
 * locally. A KST evening is the same ST afternoon, but anything after 11:00
 * KST has already rolled over — so the date has to be shown, not just the
 * time, or "13:30 ST" beside "Sun 00:30" looks like a mistake.
 */
export function crossesServerDay(iso) {
  const d = new Date(iso)
  const local = new Intl.DateTimeFormat('en-CA', { year: 'numeric', month: '2-digit', day: '2-digit' })
  const server = new Intl.DateTimeFormat('en-CA', opts({ year: 'numeric', month: '2-digit', day: '2-digit' }))
  return local.format(d) !== server.format(d)
}

/** The pairing used throughout the UI: local time, with ST alongside. */
export const withServerTime = (iso) =>
  crossesServerDay(iso) ? fmtServerFull(iso) : fmtServerTime(iso)

File: backend/src/dwsbot/servertime.py
"""Dark War Survival server time.

The game runs on a single clock that every server event is announced against,
two hours behind UTC. The alliance's anchor for it is 00:00 ST = 11:00 KST,
which is the same statement: KST is UTC+9, so ST is UTC+9-11 = UTC-2.

POSIX inverts the sign in the Etc zones, so "Etc/GMT+2" *is* UTC-02:00. It has
no daylight saving, which matches the game.
"""
from __future__ import annotations

from datetime import datetime
from zoneinfo import ZoneInfo

SERVER_TZ_NAME = "Etc/GMT+2"
SERVER_TZ = ZoneInfo(SERVER_TZ_NAME)

#: Shown wherever the zone is offered as a choice.
SERVER_TZ_LABEL = "Game server time (ST)"


def to_server(moment: datetime) -> datetime:
    """Convert any aware datetime into server time."""
    return moment.astimezone(SERVER_TZ)


def format_server(moment: datetime, *, with_date: bool = True) -> str:
    """Render a moment as the alliance would say it: "13:30 ST" or with a date."""
    st = to_server(moment)
    return st.strftime("%d %b %H:%M ST") if with_date else st.strftime("%H:%M ST")
